amazon/ion/timestamp.c

```diff
@@ -0,0 +1,290 @@
+
+   if (!PyArg_ParseTupleAndKeywords(args, kwds, "OOOOOO|OOO", keywords, &year, &month, &day, &hour, &minutes, &seconds, &useconds, &off_hours, &off_minutes)) {
+      printf("ERROR on parse tuple...\n");
+      PyErr_Print();
+      return NULL;
+   }
+
+   int precision = -1;
+   if (year != Py_None) {
+      precision++;
+   } else {
+      PyErr_Format(PyExc_ValueError, "Cannot create a timestamp with lower precision than Year.");
+      return NULL;
+   }
+   if (month != Py_None) { precision++; }
+   if (day != Py_None) { precision++; }
+   if (minutes != Py_None) { precision++; }
+   if (seconds != Py_None) { precision++; }
+   if (useconds != Py_None) {
+      Py_INCREF(useconds); // Incref because we're going to store it in our timestamp object.
+      precision++;
+      if (PyObject_IsInstance(useconds, _decimal_constructor)) {
+
+         // If our microseconds contain fractional component, we need to remove it for our parent datetime.
+         PyObject *tuple_func = PyUnicode_FromString("as_tuple");
+         PyObject *tuple = PyObject_CallMethodNoArgs(useconds, tuple_func);
+         if (PyErr_Occurred()) {
+            if (tuple_func != Py_None)
+               Py_XDECREF(tuple_func);
+            if (tuple != Py_None)
+               Py_XDECREF(tuple);
+            goto fail;
+         }
+
+         PyObject *exp = PyObject_GetAttrString(tuple, "exponent"); // Will be negative; New Ref
+         PyObject *digits = PyObject_GetAttrString(tuple, "digits"); // New Ref
+         int exponent = PyLong_AsLong(exp);
+         Py_ssize_t end = PyTuple_Size(digits) + exponent;
+
+         uint32_t usec_int = 0;
+         for (int i=0; i < end+exponent; i++) {
+            PyObject *digit = PyTuple_GetItem(digits, i); // Borrowed
+            uint32_t digit_int = PyLong_AsLong(digit);
+            usec_int += digit_int * pow(10, end - i - 1);
+         }
+         useconds_whole = PyLong_FromLong(usec_int);
+
+         Py_DECREF(exp);
+         Py_DECREF(digits);
+         Py_DECREF(tuple_func);
+         Py_DECREF(tuple);
+      } else {
+         useconds_whole = useconds;
+         Py_INCREF(useconds_whole); // Increment this, so that we can decrement later for both paths.
+      }
+   }
+
+   PyObject *datetime_args = PyTuple_Pack(7, year, month, day, hour, minutes, seconds, useconds_whole);
+   IonTimestamp *self = (IonTimestamp *) (tpe->tp_base->tp_new(tpe, datetime_args, NULL)); // DateTime __new__
+   if (self == NULL || PyErr_Occurred()) {
+      Py_XDECREF(self);
+      goto fail;
+   }
+
+   self->precision = precision;
+   self->microseconds = useconds;
+fail:
+   if (useconds_whole != Py_None)
+      Py_XDECREF(useconds_whole);
+   Py_XDECREF(datetime_args);
+   return (PyObject *)self;
+}
+
+static PyObject *IonTimestamp_FromComponents(PyObject *year, PyObject *month, PyObject *day, PyObject *hour, PyObject *min, PyObject *sec, PyObject *usec, PyObject *off_hour, PyObject *off_min) {
```

Comment:
Would `Timestamp_FromComponents` be more accurate in this case? Same as `IonTimestamp_new`, would `Timestamp_New` would be more accurate since both these methods are constructing a PyObject.